iac-extensions/docker/src/main/java/org/sonar/iac/docker/utils/ArgumentUtils.java

```diff
@@ -38,68 +38,79 @@
+        return ArgumentResolution.EMPTY;
       }
-      return new ArgumentResolution(sb.toString());
+      resolveExpressions(argument.expressions());
+      return resolution.build();
     }
 
-    @Nullable
-    private String resolveExpression(Expression expression) {
-      if (expression.is(DockerTree.Kind.STRING_LITERAL)) {
-        return ((Literal)expression).value();
-      }
-      if (expression.is(DockerTree.Kind.EXPANDABLE_STRING_LITERAL)) {
-        return resolveExpressions(((ExpandableStringLiteral)expression).expressions()).value();
-      }
-      if (expression.is(DockerTree.Kind.EXPANDABLE_STRING_CHARACTERS)) {
-        return ((ExpandableStringCharacters)expression).value();
-      }
-      if (expression.is(DockerTree.Kind.REGULAR_VARIABLE)
-        || (expression.is(DockerTree.Kind.ENCAPSULATED_VARIABLE) && !":+".equals(((EncapsulatedVariable) expression).modifierSeparator()))) {
-        return resolveVariable((Variable) expression).value();
+    private void resolveExpressions(List<Expression> expressions) {
+      for (Expression expression : expressions) {
+        resolveExpression(expression);
       }
-      return null;
     }
 
-    private ArgumentResolution resolveVariable(Variable variable) {
-      Symbol symbol = variable.symbol();
-      return symbol != null ? resolveSymbol(symbol) : ArgumentResolution.EMPTY;
+    private void resolveExpression(Expression expression) {
+      switch (expression.getKind()) {
+        case STRING_LITERAL:
+          resolution.addValue(((Literal)expression).value());
+          break;
+        case EXPANDABLE_STRING_CHARACTERS:
+          resolution.addValue(((ExpandableStringCharacters)expression).value());
+          break;
+        case EXPANDABLE_STRING_LITERAL:
+          resolveExpressions(((ExpandableStringLiteral)expression).expressions());
+          break;
+        case REGULAR_VARIABLE:
+          resolveVariable((Variable) expression);
+          break;
+        case ENCAPSULATED_VARIABLE:
+          EncapsulatedVariable encapsulatedVariable = (EncapsulatedVariable) expression;
+          if (!":+".equals((encapsulatedVariable).modifierSeparator())) {
+            resolveVariable(encapsulatedVariable);
+          } else {
+            resolution.addPlaceholder(encapsulatedVariable);
+          }
+          break;
+        default:
+      }
     }
 
     /**
      * To resolve the value of a symbol at a given state, the last assigned value is considered.
      * There for all symbol usages are analyzed for the last assignment with value before the access.
      */
-    private ArgumentResolution resolveSymbol(Symbol symbol) {
+    private void resolveVariable(Variable variable) {
+      Symbol symbol = variable.symbol();
+      if (!visitedVariable.add(variable) || symbol == null) {
+        resolution.addPlaceholder(variable);
+        return;
+      }
+
       List<Usage> usages = symbol.usages();
       Collections.reverse(usages);
```

Comment:
I just saw that this will create a bug.
By calling `Collection.reverse()`, it reverse the elements of the provided list.
Since `symbol.usages()` provide the original list of usages, everytime we resolve a variable, we reverse the order of usages for the symbol.
This is easily testable in `ArgumentUtilsTest.shouldResolveLabelValue()`, by doing a second resolution :
```
ArgumentUtils.ArgumentResolution resolution = ArgumentUtils.resolve(label.value());
assertThat(resolution.value()).isEqualTo("bar");

ArgumentUtils.ArgumentResolution resolution2 = ArgumentUtils.resolve(label.value());
assertThat(resolution2.value()).isEqualTo("bar");
```
A possible solution could be to use `Lists.reverse()`, which return a view and does not modify the original List (and also is an operation which take constant time instead of linear time).

Reply:
Ah, nice catch. I had this issue this week when playing around with the implementation. I will fix it.